conveyer/queue/mutex_queue.go

```diff
@@ -0,0 +1,224 @@
+	newNode.index = q.head.index + 1
+
+	newNode.biggestQueueSignal = maxSignal(q.head.biggestQueueSignal, newNode.itemType)
+
+	q.head = newNode
+
+	return nil
+}
+
+// SinkPush is implementation for IQueue
+func (q *MutexQueue) SinkPush(data interface{}) error {
+
+	newNode := &queueItem{
+		payload: data,
+	}
+
+	return q.sinkPush(newNode)
+}
+
+// SinkPushAll is implementation for IQueue
+func (q *MutexQueue) SinkPushAll(data []interface{}) error {
+	inputSize := len(data)
+	lastElement := &queueItem{}
+	newHead := lastElement
+
+	lastElement.payload = data[inputSize-1]
+	for i := inputSize - 2; i >= 0; i-- {
+		lastElement.next = &queueItem{}
+		lastElement = lastElement.next
+		lastElement.payload = data[i]
+	}
+
+	q.locker.Lock()
+	defer q.locker.Unlock()
+
+	if q.isQueueBlockedUnsafe() {
+		return errors.New("[ SinkPushAll ] Queue is blocked")
+	}
+
+	if q.HasSignal() {
+		// do smth interesting
+	}
+
+	nextIndex := q.head.index + 1
+	tmpHead := newHead
+	currentMaxSignal := maxSignal(q.head.biggestQueueSignal, tmpHead.itemType)
+	for i := inputSize - 1; i >= 0; i-- {
+		tmpHead.index = nextIndex + uint(i)
+		currentMaxSignal = maxSignal(currentMaxSignal, tmpHead.biggestQueueSignal)
+		tmpHead.biggestQueueSignal = currentMaxSignal
+		tmpHead = tmpHead.next
+	}
+
+	lastElement.next = q.head
+
+	q.head = newHead
+
+	return nil
+}
+
+func (q *MutexQueue) checkAndGetHeadUnsafe() *queueItem {
+	if q.isQueueBlockedUnsafe() || q.isQueueEmptyUnsafe() {
+		return nil
+	}
+
+	return q.head
+}
+
+func (q *MutexQueue) isQueueBlockedUnsafe() bool {
+	return q.head == nil
+}
+
+func (q *MutexQueue) isQueueEmptyUnsafe() bool {
+	return q.head == &emptyQueueItem
+}
+
+// get pointer to head and unfold linked list to slice:
+//  all signals will be at the begging of the slice
+func convertSublistToArray(localHead *queueItem) []OutputElement {
+	result := make([]OutputElement, localHead.index)
+
+	current := localHead
+	signalCurrentIndex := 0
+	messageCurrentIndex := localHead.index - 1
+	for i := uint(0); i < localHead.index; i++ {
+		element := OutputElement{
+			data:     current.payload,
+			itemType: current.itemType,
+		}
+		if current.isSignal() {
+			result[signalCurrentIndex] = element
+			signalCurrentIndex++
+		} else {
+			result[messageCurrentIndex] = element
+			messageCurrentIndex--
+		}
+		current = current.next
+	}
+
+	return result
+}
+
+// SinkPushAll is implementation for IQueue
+func (q *MutexQueue) RemoveAll() []OutputElement {
+
+	var localHead *queueItem
+	q.locker.Lock()
+	localHead = q.checkAndGetHeadUnsafe()
+	if localHead == nil {
+		q.locker.Unlock()
+		return []OutputElement{}
+	}
+	q.head = &emptyQueueItem
+	q.locker.Unlock()
+
+	return convertSublistToArray(localHead)
+}
+
+func (q *MutexQueue) BlockAndRemoveAll() []OutputElement {
+	var localHead *queueItem
+	q.locker.Lock()
+	localHead = q.checkAndGetHeadUnsafe()
+	if localHead == nil {
+		q.head = nil
+		q.locker.Unlock()
+		return []OutputElement{}
+	}
+	q.head = nil
```

Comment:
Why not do `q.head = nil` and `q.locker.Unlock()` before `if` condition?

Reply:
Since localHead is pointer to q.head, and if I set q.head = nil before if statement - this condition will be true every time